### PR TITLE
More fuzzy matching and matching multiple words for window switcher

### PR DESCRIPTION
**Pull request by hugovalk**
The Window Switcher plugin is quite useful to me, but I had trouble matching some of the windows on my desktop. This happens because the wm_class is not always descriptive enough. For example:

- All Java apps have a wm_class that starts with: `sun-awt-X11-XFramePeer.`. 
- Also sometimes the first segment of the wm_class is something generic, like for example `main.terminology` for the Terminology terminal app. 

In all those cases, the 'nice' application name can be found after the dot in wm_class or in the wm_name field.

This change matches all words in the query (splitted on whitespace) with wm_class OR wm_name. This enables me switch to all windows on my screen. Also, it allows switching to the right IntelliJ project, since the project name happens to be in the wm_name. 
I have tested this change on Arch Linux with KDE and Ubuntu 18.04 with Gnome.

### CODE REVIEW

File: window_switcher.py [hunk 1]
```diff
@@ -14,20 +14,23 @@
...
-__version__ = "1.4"
+__version__ = "1.5"
 __author__ = "Ed Perez, Manuel Schneider"
 __dependencies__ = ["wmctrl"]
 
 if which("wmctrl") is None:
     raise Exception("'wmctrl' is not in $PATH.")
 
+def matchWindow(win, text):
+    return (text in win.wm_class.lower() or text in win.wm_name.lower())
```
Comment (ManuelSchneid3r) [inline window_switcher.py]: please no single line functions. This does not really improve maintainability/readability.

File: window_switcher.py [hunk 2]
```diff
@@ -14,20 +14,23 @@
...
+
 def handleQuery(query):
-    stripped = query.string.strip().lower()
+    stripped = query.string.strip().lower().split()
     if stripped:
         results = []
         for line in subprocess.check_output(['wmctrl', '-l', '-x']).splitlines():
             win = Window(*[token.decode() for token in line.split(None,4)])
-            if win.desktop != "-1"  and stripped in win.wm_class.split('.')[0].lower():
+            if win.desktop != "-1" and all(matchWindow(win, text) for text in stripped):
```
Comment (ManuelSchneid3r) [inline window_switcher.py]:
```python
if win.desktop != "-1" and all(txt in win.wm_class.lower() or txt in win.wm_name.lower() for txt in stripped)
```

Should do it as well

File: window_switcher.py [hunk 3]
```diff
@@ -14,20 +14,23 @@
...
 
 if which("wmctrl") is None:
     raise Exception("'wmctrl' is not in $PATH.")
 
+def matchWindow(win, text):
+    return (text in win.wm_class.lower() or text in win.wm_name.lower())
+
 def handleQuery(query):
-    stripped = query.string.strip().lower()
+    stripped = query.string.strip().lower().split()
```
Comment (ManuelSchneid3r) [inline window_switcher.py]: This does not handle consecutive spaces. `.split()` does sth similar "strip" anyway. Use
```python
filter(None, query.string.lower().split())
```

File: window_switcher.py [hunk 4]
```diff
@@ -14,20 +14,30 @@
 
 __iid__ = "PythonInterface/v0.1"
 __prettyname__ = "Window Switcher"
-__version__ = "1.4"
+__version__ = "1.5"
 __author__ = "Ed Perez, Manuel Schneider"
```
Comment (ManuelSchneid3r) [inline window_switcher.py]: btw add your name here

**Comment by vthuongt** [comment]
I edited the the script and added some more nice features such as tokenized + fuzzy search and more. However I don't know how to add my code to this pull request. You can check out the code here:
https://github.com/vthuongt/window-switcher-plus

Thank you for the cool launcher!

**Comment by hugovalk** [comment]
I have still a question open on this PR. Any chance it could be answered? @ManuelSchneid3r

**Comment by ManuelSchneid3r** [comment]
tely I let you all wait pretty long. Now I have the problem that I don't know which PR's are valid at all. Therefore I will close all PR's for tentatively. If you want me to reopen the PR let me know. Also it would be nice if you could resolve potential conflicts due to changes in the repo. Sorry for the inconveniences, but I am working on it for a week now. I'm in a hurry. Note that this is a copy paste message. 😉

**Pull request closed without merging**